Tests/WorkspaceTests/WorkspaceTests.swift

```diff
@@ -559,6 +559,130 @@ final class WorkspaceTests: XCTestCase {
+                rootDeps: [
+                    MockDependency("A", version: Version(1, 0, 0)..<Version(1, .max, .max)),
+                ],
+                packages: [
+                    MockPackage("A", version: v1)
+                ]
+            )
+            
+            let delegate = TestWorkspaceDelegate()
+            let repoPath = AbsolutePath(manifestGraph.repo("A").url)
+            
+            func createWorkspace(with manifestGraph: MockManifestGraph) throws -> Workspace {
+                return  Workspace.createWith(rootPackage: path, manifestLoader: manifestGraph.manifestLoader, delegate: delegate)
+            }
+            
+            do {
+                var workspace = try createWorkspace(with: manifestGraph)
+                
+                // Ensure delegates haven't been called yet.
+                XCTAssert(delegate.fetched.isEmpty)
+                XCTAssert(delegate.cloned.isEmpty)
+                XCTAssert(delegate.checkedOut.isEmpty)
+                XCTAssert(delegate.removed.isEmpty)
+                
+                // Load the package graph.
+                let diagnostics = DiagnosticsEngine()
+                let graph = workspace.loadPackageGraph(rootPackages: [path], diagnostics: diagnostics)
+                
+                XCTAssertFalse(diagnostics.hasErrors)
+                
+                // Test the delegates.
+                XCTAssert(delegate.fetched.count == 1)
+                XCTAssert(delegate.cloned.count == 1)
+                XCTAssert(delegate.removed.isEmpty)
+                
+                for (_, repoPath) in manifestGraph.repos {
+                    XCTAssert(delegate.fetched.contains(repoPath.url))
+                    XCTAssert(delegate.cloned.contains(repoPath.url))
+                    XCTAssertEqual(delegate.checkedOut[repoPath.url], "1.0.0")
+                }
+                
+                // Validate that the graph has the correct basic structure.
+                XCTAssertEqual(graph.packages.count, 2)
+                XCTAssertEqual(graph.packages.map{ $0.name }.sorted(), ["A", "Root"])
+                
+                
+                let rootManifests = workspace.loadRootManifests(packages: [path], diagnostics: diagnostics)
+                let manifests = workspace.loadDependencyManifests(rootManifests: rootManifests, diagnostics: diagnostics)
+                
+                XCTAssertFalse(diagnostics.hasErrors)
+                
+                guard let aManifest = manifests.lookup(manifest: "A") else {
+                    return XCTFail("Expected manifest for package A not found")
+                }
+                
+                func getDependency(_ manifest: Manifest) -> ManagedDependency {
+                    return workspace.managedDependencies[manifest.url]!
+                }
+                
+                let dependency = getDependency(aManifest)
+        
```

Comment:
We can avoid all these asserts above because they are already covered by other tests.